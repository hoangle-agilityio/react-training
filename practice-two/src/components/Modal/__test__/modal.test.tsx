import { cleanup, fireEvent, render } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import ModalUser from "..";

// Create mock for modal
jest.mock("../index");

// Create mock for api
jest.mock("../../../services/index");

window.alert = jest.fn();

const modalProps = {
  open: true,
  isViewUser: false,
  onEdit: jest.fn(),
  closeModal: jest.fn(),
}

const currentUser = {
  id: 1,
  name: "Hoang",
  email: "hoang.le@asnet.com.vn"
}

describe('Modal Component', () => {
  afterEach(cleanup);

  it('should be render modal title as Add User', () => {
    const { container } = render(<ModalUser {...modalProps} />);

    expect(container.querySelector(".modal-title")?.textContent).toBe("Add User");
    expect(container.querySelector(".modal-title")).toMatchSnapshot();
  });

  it('should be render modal title as Update User', () => {
    const { container } = render(<ModalUser {...modalProps} currentUser={currentUser} />);

    expect(container.querySelector(".modal-title")?.textContent).toBe("Update User");
    expect(container.querySelector(".modal-title")).toMatchSnapshot();
  });

  it('should be render modal title as View User', () => {
    const { container } = render(<ModalUser {...modalProps} isViewUser={true} currentUser={currentUser} />);

    expect(container.querySelector(".modal-title")?.textContent).toBe("View User");
    expect(container.querySelector(".modal-title")).toMatchSnapshot();
  });

  it('should be button click event to move to edit modal', () => {
    const { getByRole } = render(<ModalUser {...modalProps} isViewUser={true} currentUser={currentUser} />);
    const button = getByRole("button", { name: /Edit/i });
    const onClick = fireEvent.click(button);

    expect(onClick).toBe(true);
    expect(button).toMatchSnapshot();
  });

  it('should be button click event to update user', () => {
    const { getByRole } = render(<ModalUser {...modalProps} currentUser={currentUser} />);
    const button = getByRole("button", { name: /Update User/i });
    const onClick = fireEvent.click(button);

    expect(onClick).toBe(true);
    expect(button).toMatchSnapshot();
  });

  it('should be button click event to create user', () => {
    const { getByPlaceholderText, getByRole } = render(<ModalUser {...modalProps} />);
    const userName = getByPlaceholderText(/user name/i);
    const userEmail = getByPlaceholderText(/user email/i);
    const button = getByRole("button", { name: /Add User/i });

    userEvent.type(userName, "Hoang Le");
    userEvent.type(userEmail, "hoang.le@gmail.com");

    const onClick = fireEvent.click(button);

    expect(onClick).toBe(true);
    expect(button).toMatchSnapshot();
  });
});
