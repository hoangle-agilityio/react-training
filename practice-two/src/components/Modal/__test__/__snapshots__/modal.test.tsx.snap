// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Modal Component should be button click event to create user 1`] = `
<button
  class="btn btn-success"
>
  Add User
</button>
`;

exports[`Modal Component should be button click event to move to edit modal 1`] = `
<button
  class="btn btn-primary"
>
  Edit
</button>
`;

exports[`Modal Component should be button click event to update user 1`] = `
<button
  class="btn btn-success"
>
  Update User
</button>
`;

exports[`Modal Component should be render modal title as Add User 1`] = `
<h2
  class="modal-title"
>
  Add User
</h2>
`;

exports[`Modal Component should be render modal title as Update User 1`] = `
<h2
  class="modal-title"
>
  Update User
</h2>
`;

exports[`Modal Component should be render modal title as View User 1`] = `
<h2
  class="modal-title"
>
  View User
</h2>
`;
